sdk/messaging/azservicebus/message.go

```diff
@@ -4,105 +4,331 @@
+		linkRevision uint64
+	}
+
+	// Message is a SendableMessage which can be sent using a Client.NewSender().
+	Message struct {
+		ID string
+
+		ContentType   string
+		CorrelationID string
+		// Body corresponds to the first []byte array in the Data section of an AMQP message.
+		Body             []byte
+		SessionID        *string
+		Subject          string
+		ReplyTo          string
+		ReplyToSessionID string
+		To               string
+		TimeToLive       *time.Duration
+
+		PartitionKey            *string
+		TransactionPartitionKey *string
+		ScheduledEnqueueTime    *time.Time
+
+		ApplicationProperties map[string]interface{}
+		Format                uint32
+	}
 )
 
-// Message represents a sendable message, including only user-settable properties.
-// A message that has been received from Service Bus will be of type `ReceivedMessage`.
-type Message struct {
-	Body                    []byte
-	ID                      string
-	PartitionKey            *string
-	TransactionPartitionKey *string
+// Service Bus custom properties
+const (
+	// DeliveryAnnotation properties
+	lockTokenDeliveryAnnotation = "x-opt-lock-token"
+
+	// Annotation properties
+	partitionKeyAnnotation           = "x-opt-partition-key"
+	viaPartitionKeyAnnotation        = "x-opt-via-partition-key"
+	scheduledEnqueuedTimeAnnotation  = "x-opt-scheduled-enqueue-time"
+	lockedUntilAnnotation            = "x-opt-locked-until"
+	sequenceNumberAnnotation         = "x-opt-sequence-number"
+	enqueuedTimeAnnotation           = "x-opt-enqueued-time"
+	deadLetterSourceAnnotation       = "x-opt-deadletter-source"
+	enqueuedSequenceNumberAnnotation = "x-opt-enqueue-sequence-number"
+)
 
-	// TODO: is an empty session ID valid?
-	SessionID string
+// Set implements tab.Carrier
+func (m *Message) Set(key string, value interface{}) {
+	if m.ApplicationProperties == nil {
+		m.ApplicationProperties = make(map[string]interface{})
+	}
+	m.ApplicationProperties[key] = value
 }
 
-// ReceivedMessage represents a message received from Service Bus.
-type ReceivedMessage struct {
-	Message
-	LockToken      string
-	SequenceNumber int64
+// GetKeyValues implements tab.Carrier
+func (m *Message) GetKeyValues() map[string]interface{} {
+	return m.ApplicationProperties
+}
 
-	legacyMessage *internal.Message
+func (m *Message) MessageType() string {
+	return "Message"
 }
 
-func (m *Message) toAMQPMessage() (*amqp.Message, error) {
-	messageID := m.ID
+func (m *Message) ToAMQPMessage() *amqp.Message {
+	amqpMsg := amqp.NewMessage(m.Body)
+
+	if m.TimeToLive != nil {
+		if amqpMsg.Header == nil {
+			amqpMsg.Header = new(amqp.MessageHeader)
+		}
+		amqpMsg.Header.TTL = *m.TimeToLive
+	}
 
 	// TODO: I don't think this should be strictly required. Need to
 	// look into why it won't send properly without one.
+	var messageID = m.ID
+
 	if messageID == "" {
 		uuid, err := uuid.NewV4()
 
-		if err != nil {
-			return nil, err
+		if err == nil {
```

Comment:
What happens in the error case?

Reply:
They just end up with an empty message ID.

There's no good default to give it here, unfortunately.